test/test_factbase.rb

```diff
@@ -219,4 +220,150 @@ def test_txn_with_rollback
+    end
+    assert_equal(100, fb.size)
+    assert_equal(100, fb.query('(eq foo 42)').each.to_a.size)
+    assert_equal(100, fb.query('(eq bar 49)').each.to_a.size)
+    assert_equal(100, fb.query("(eq value #{42 * 49})").each.to_a.size)
+  end
+
+  def test_different_values_when_concurrent_inserts
+    fb = Factbase.new
+    n = 1
+    Threads.new(100).assert do
+      fb.insert.foo = n
+      n += 1
+    end
+    assert_equal(100, fb.size)
+    100.times do |num|
+      i = num + 1
+      f = fb.query("(eq foo #{i})").each.to_a
+      assert_equal(1, f.count)
+      assert_equal(i, f.first.foo)
+    end
+  end
+
+  def test_different_properties_when_concurrent_inserts
+    fb = Factbase.new
+    n = 1
+    Threads.new(100).assert do
+      fb.insert.send(:"prop_#{n}=", n)
+      n += 1
+    end
+    assert_equal(100, fb.size)
+    100.times do |num|
+      i = num + 1
+      prop = "prop_#{i}"
+      f = fb.query("(eq #{prop} #{i})").each.to_a
+      assert_equal(1, f.count)
+      assert_equal(i, f.first.send(prop.to_sym))
+    end
+  end
+
+  def test_concurrent_transactions_inserts
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        i += 1
+      end
+    end
+    assert_equal(100, fb.size)
+    assert_equal(100, fb.query('(exists thread_id)').each.to_a.size)
+  end
+
+  def test_concurrent_transactions_with_rollbacks
+    fb = Factbase.new
+    Threads.new(100).assert do |i|
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        raise Factbase::Rollback
+      end
+    end
+    assert_equal(0, fb.size)
+  end
+
+  def test_concurrent_transactions_successful
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fb.txn do |fbt|
+        fact = fbt.insert
+        fact.thread_id = i
+        fact.value = i * 10
+        i += 1
+      end
+    end
+    facts = fb.query('(exists thread_id)').each.to_a
+    assert_equal(100, facts.size)
+    facts.each do |fact|
+      assert_equal(fact.value, fact.thread_id * 10)
+    end
+  end
+
+  def test_concurrent_queries
+    fb = Factbase.new
+    i = 0
+    Threads.new(100).assert do
+      fact = fb.insert
+      fact.thread_id = i
+      fact.value = i * 10
+      i += 1
+    end
+    Threads.new(100).assert do
+      results = fb.query('(exists thread_id)').each.to_a
+      assert_equal(100, results.size)
+
+      thread_ids = results.map(&:thread_id)
+      assert_equal((0..99).to_a, thread_ids.sort)
+    end
+  end
+
+  def test_export_import_concurrent
+    fb = Factbase.new
+    Threads.new(100).assert do
+      fact = fb.insert
+      fact.foo = 42
+    end
+    exported_data = nil
+    Threads.new(1).assert do
+      exported_data = fb.export
+    end
+    Threads.new(5).assert do
+      new_fb = Factbase.new
+      new_fb.import(exported_data)
+      assert_equal(fb.size, new_fb.size)
+      fb.query('(exists thread_id)').each.to_a do |fact|
```

Comment:
The block will never be called if this query returns an empty list. I would suggest adding an extra assert. Smth like

```ruby
assert fb.query('(exists thread_id)').each.any?
```

or

```ruby
facts = fb.query('(exists thread_id)').each.to_a
assert_equal(100,  facts.size)
```

Like you do this in other tests